lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+        Attribute::OBJECT_FIT,
+        Attribute::OBJECT_POSITION,
+        Attribute::REFERRERPOLICY,
+        Attribute::SRC,
+        Attribute::SRCSET,
+        Attribute::SIZES,
+        Attribute::TITLE,
+    ];
+
+    /**
+     * Images smaller than 150px are considered tiny.
+     *
+     * @var int
+     */
+    const TINY_IMG_THRESHOLD = 150;
+
+    /**
+     * Maximum number of hero images defined via data-hero attribute.
+     *
+     * @var int
+     */
+    const DATA_HERO_MAX = 2;
+
+    /**
+     * Configuration store to use.
+     *
+     * @var TransformerConfiguration
+     */
+    private $configuration;
+
+    /**
+     * Instantiate a PreloadHeroImage object.
+     *
+     * @param TransformerConfiguration $configuration Configuration store to use.
+     */
+    public function __construct(TransformerConfiguration $configuration)
+    {
+        $this->configuration = $configuration;
+    }
+
+    /**
+     * Apply transformations to the provided DOM document.
+     *
+     * @param Document        $document DOM document to apply the transformations to.
+     * @param ErrorCollection $errors   Collection of errors that are collected during transformation.
+     * @return void
+     */
+    public function transform(Document $document, ErrorCollection $errors)
+    {
+        if ($this->configuration->get(PreloadHeroImageConfiguration::PRELOAD_HERO_IMAGE) === false) {
+            return;
+        }
+
+        $heroImages    = $this->findHeroImages($document);
+        $referenceNode = $document->viewport;
+
+        $heroImageCount = count($heroImages);
+        if ($heroImageCount > self::DATA_HERO_MAX) {
+            $errors->add(Error\TooManyHeroImages::whenPastMaximum());
+            $heroImageCount = self::DATA_HERO_MAX;
+        }
+
+        $isAmpStory = Amp::isAmpStory($document);
+
+        for ($index = 0; $index < $heroImageCount; $index++) {
+            $this->generatePreload($heroImages[$index], $document, $errors, $referenceNode);
+            if (! $isAmpStory) {
+                // AMP Stories don't support SSR'd <amp-img> yet.
+                // See https://github.com/ampproject/amphtml/issues/29850.
+                $this->generateImg($heroImages[$index], $document);
+            }
+        }
+    }
+
+    /**
+     * Find the hero images to optimize.
+     *
+     * @param Document $document Document to look for hero images in.
+     * @return HeroImage[] Array of hero images to optimize.
+     */
+    private function findHeroImages(Document $document)
+    {
+        $heroImageCandidate = null;
+        $heroImages         = [];
+        $node               = $document->body;
+
+        while ($node !== null) {
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
```

Comment:
If `count( $heroImages ) === self::DATA_HERO_MAX`, this could `return $heroImages` as opposed to continuing to iterate over the rest of the document, no? 

Or is it preferable to explicitly emit the error?

Reply:
I think it is preferable to collect a list of all offending images for the purpose or diagnosing/correcting the issue.